### ISSUE TITLE
Question: Plans for multi-user support?

### ISSUE THREAD

**Issue by WorldWideWebb** [caprover/caprover#492]
Any plans on introducing user management? Would love to start rolling this out, but since we have multiple developers here, we would need to be able to control access to different apps/repositories.

**Comment by sean-nicholas** [comment]
Would love to have this feature, too. I am willing to contribute as well :)

**Comment by githubsaturn** [comment]
TLDR: No plan in near future. 

More details:
Multi-user support is way more complicated that it sounds. Define roles and actions are complex and cause confusion and frustration and security risks if the implementation is not perfect. Due to the very nature of CapRover, it's not something that we can gradually implement and go with piece by piece implementation.
To shed more light on how complex it can be, assume that we have a role that can create apps, and the other role that can modify certain apps (binding ports, building, and etc). Theoretically, if we say we support these two roles, one assumes that the second role would be limited and cannot access other apps. Wrong! Someone with access level of role 2 can bind to the Docker unix socket port, and completely take over the host machine - even be able to delete, control, or read sensitive data of other apps. Another role-2 level user can deploy a build that can take down the entire server due to high memory consumption. There are numerous other problems that a multi-user PaaS has to address that makes this "feature" ridiculously difficult. A "potential" approach that some PaaS solutions take is to limit the PaaS to a very small set of features, such that there is no support for volume mapping, port binding, custom nginx config and etc. Of course, CapRover has taken the exact opposite of this approach which is to let the users to customize the behavior if they need to.

**Comment by planecore** [comment]
I fully agree with @lcx. This is a fairly basic feature and it should be available.

**Comment by azophy** [comment]
A little suggestion, instead of implementing Authentication and multi-user Authorization from scratch, we could leverage established OIDC-based auth management solutions such as Okta or Keycloak. This approach at least should reduce the feature that need to be implemented in the future. Another open-source project I know that use this approach is [Outline](https://github.com/outline/outline)

**Comment by kasvith** [comment]
There is also authz libs like Casl and Casbin in such case.
At least heroku level multi user support should be there

**Comment by SaadBazaz** [comment]
If we could just have user roles called "admin" (can overlook all container) and "viewer" (can just see metrics / deployment logs / start stop their containers like Vercel) that'd be amazing. And it seems simple enough. The "admin" can be the current user which already works, the "viewer" just needs restricted permissions. 

Nothing more nothing less. This can take Caprover to another level.

**Comment by mohamedhagag** [comment]
As per @azophy comment it will be good to implement multi-user with same admin permissions using OIDC

Also we can look at how other projects like portainer implement it in the open source version
Multi-user is complex, but it have been done lots of times before in similar FOSS in Virtualization management to containers and multi hypervisor solutions like Portainer, Proxmox , OpenNebula, oVirt ... etc .

**Comment by isaccavalcante** [comment]
@SaadBazaz idea would be great. Specially for separating production from development environments.
